src/main/java/org/opentripplanner/model/TripPattern.java

```diff
@@ -734,4 +718,41 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
     // to ensure that the stops field it uses in TripPattern is already deserialized.
     scheduledTimetable.finish();
   }
+
+  /**
+   * Check if given stop and next stop on this trip pattern and other are equal.
+   *
+   * @param other Other instance of trip pattern with list of stops. May not be null.
+   * @param index Given index for stop
+   * @return true if stop and next stop are equal on bouth trip patterns, else false
+   */
+  private boolean sameStops(TripPattern other, int index) {
+    var otherOrigin = other.getStop(index);
+    var otherDestination = other.getStop(index + 1);
+    var origin = getStop(index);
+    var destination = getStop(index + 1);
+
+    return origin.equals(otherOrigin) && destination.equals(otherDestination);
+  }
+
+  /**
+   * Check if Station is equal on given stop and next stop for this trip pattern and other.
+   *
+   * @param other Other instance of trip pattern with list of stops. May not be null.
+   * @param index Given index for stop
+   * @return true if the stops have the same stations, else false. If any station is null then
+   * false.
+   */
+  private boolean sameStations(TripPattern other, int index) {
+    var otherOrigin = other.getStop(index).getParentStation();
+    var otherDestination = other.getStop(index + 1).getParentStation();
+    var origin = getStop(index).getParentStation();
+    var destionation = getStop(index + 1).getParentStation();
+
+    if (origin == null || destionation == null) {
```

Comment:
What if only one of these is is a station, while the other one is a stop, that is equal?

Reply:
Not sure if I understand what you mean. Isn't Trip Pattern a collection of stops by definition?

Reply:
> You mean that getParentStation() may return null

Reply:
> You mean that getParentStation() may return null

Exactly. Consider that we have a timetable of (stop, station) tuples, which is (A1, A) -> (B, null) and a realtime update (A2, A) -> (B, null). Or the opposite where timetable is (A, null) -> (B1, B) and realtime is (A, null) -> (B2, B)

Reply:
If station is null then I consider stations equal if stops are equal